From 388331975b35d39867d8c1db74b7e3769c1fe1d1 Mon Sep 17 00:00:00 2001
From: Roman Stratiienko <r.stratiienko@gmail.com>
Date: Sun, 19 Nov 2023 19:26:36 +0200
Subject: [PATCH 1/5] Add init_boot partition

[1] states: "init_boot image added for devices launching with Android 13"

test: atest GenericBootImageTest

[1]: https://source.android.com/docs/core/architecture/partitions/generic-boot
Signed-off-by: Roman Stratiienko <r.stratiienko@gmail.com>
---
 common/base/board.mk              | 13 +++++++++++--
 platform/tools/flash-all.sh       |  1 +
 platform/tools/gensdimg.sh        |  2 ++
 platform/tools/tools.mk           |  1 +
 platform/uboot/bootscript.cpp     |  8 +++++++-
 platform/uboot/platform_amlogic.h |  3 ++-
 6 files changed, 24 insertions(+), 4 deletions(-)

diff --git a/common/base/board.mk b/common/base/board.mk
index 3ca17a5..19aa1ff 100644
--- a/common/base/board.mk
+++ b/common/base/board.mk
@@ -46,11 +46,17 @@ BOARD_AVB_VBMETA_SYSTEM_ROLLBACK_INDEX_LOCATION := 1
 #BOARD_AVB_BOOT_ROLLBACK_INDEX := $(PLATFORM_SECURITY_PATCH_TIMESTAMP)
 #BOARD_AVB_BOOT_ROLLBACK_INDEX_LOCATION := 2
 
+# Enable chained vbmeta for init_boot images
+#BOARD_AVB_INIT_BOOT_KEY_PATH := external/avb/test/data/testkey_rsa2048.pem
+#BOARD_AVB_INIT_BOOT_ALGORITHM := SHA256_RSA2048
+#BOARD_AVB_INIT_BOOT_ROLLBACK_INDEX := $(PLATFORM_SECURITY_PATCH_TIMESTAMP)
+#BOARD_AVB_INIT_BOOT_ROLLBACK_INDEX_LOCATION := 3
+
 # Enable chained vbmeta for vendor_boot images
 #BOARD_AVB_VENDOR_BOOT_KEY_PATH := external/avb/test/data/testkey_rsa4096.pem
 #BOARD_AVB_VENDOR_BOOT_ALGORITHM := SHA256_RSA4096
 #BOARD_AVB_VENDOR_BOOT_ROLLBACK_INDEX := $(PLATFORM_SECURITY_PATCH_TIMESTAMP)
-#BOARD_AVB_VENDOR_BOOT_ROLLBACK_INDEX_LOCATION := 3
+#BOARD_AVB_VENDOR_BOOT_ROLLBACK_INDEX_LOCATION := 4
 
 BOARD_AVB_SYSTEM_ADD_HASHTREE_FOOTER_ARGS += --hash_algorithm sha256
 BOARD_AVB_SYSTEM_OTHER_ADD_HASHTREE_FOOTER_ARGS += --hash_algorithm sha256
@@ -89,6 +95,7 @@ BOARD_VENDORIMAGE_FILE_SYSTEM_TYPE := ext4
 BOARD_BOOTIMAGE_PARTITION_SIZE := $(shell echo $$(( 64 * 1024 * 1024 )))
 BOARD_DTBOIMG_PARTITION_SIZE := $(shell echo $$(( 512 * 1024 )))
 BOARD_VENDOR_BOOTIMAGE_PARTITION_SIZE := $(shell echo $$(( 32 * 1024 * 1024 )))
+BOARD_INIT_BOOT_IMAGE_PARTITION_SIZE := $(shell echo $$(( 8 * 1024 * 1024 )))
 
 # DKLM partition
 BOARD_USES_VENDOR_DLKMIMAGE := true
@@ -115,7 +122,7 @@ BOARD_GLODROID_DYNAMIC_PARTITIONS_SIZE := $(shell echo $$(( $(BOARD_SUPER_PARTIT
 BOARD_GLODROID_DYNAMIC_PARTITIONS_PARTITION_LIST := system system_ext vendor product vendor_dlkm
 
 AB_OTA_UPDATER := true
-AB_OTA_PARTITIONS += boot system system_ext vendor product vendor_dlkm vendor_boot vbmeta vbmeta_system dtbo
+AB_OTA_PARTITIONS += boot init_boot system system_ext vendor product vendor_dlkm vendor_boot vbmeta vbmeta_system dtbo
 
 TARGET_RECOVERY_PIXEL_FORMAT := RGBX_8888
 
@@ -124,6 +131,7 @@ BOARD_INCLUDE_RECOVERY_DTBO := true
 BOARD_INCLUDE_DTB_IN_BOOTIMG := true
 BOARD_PREBUILT_DTBIMAGE_DIR := glodroid/configuration/platform/kernel
 BOARD_BOOT_HEADER_VERSION := 4
+BOARD_INIT_BOOT_HEADER_VERSION := 4
 ifeq ($(PRODUCT_BOARD_PLATFORM),sunxi)
 BOARD_KERNEL_BASE     := 0x40000000
 BOARD_KERNEL_SRC_DIR  ?= glodroid/kernel/common-android13-5.15-lts
@@ -144,6 +152,7 @@ BOARD_KERNEL_PAGESIZE := 4096
 BOARD_KERNEL_CMDLINE  := androidboot.hardware=$(TARGET_PRODUCT)
 BOARD_MKBOOTIMG_ARGS  += --kernel_offset 0x80000 --second_offset 0x8800 --ramdisk_offset 0x3300000
 BOARD_MKBOOTIMG_ARGS  += --dtb_offset 0x3000000 --header_version $(BOARD_BOOT_HEADER_VERSION)
+BOARD_MKBOOTIMG_INIT_ARGS += --header_version $(BOARD_INIT_BOOT_HEADER_VERSION)
 
 BOARD_RAMDISK_USE_LZ4 := true
 
diff --git a/platform/tools/flash-all.sh b/platform/tools/flash-all.sh
index 2071092..9286e61 100755
--- a/platform/tools/flash-all.sh
+++ b/platform/tools/flash-all.sh
@@ -19,6 +19,7 @@ set -x
 ./fastboot flash bootloader      bootloader__SUFFIX__.img
 ./fastboot flash uboot-env       env.img
 ./fastboot flash boot            boot.img
+./fastboot flash init_boot       init_boot.img
 ./fastboot flash vendor_boot     vendor_boot.img
 ./fastboot flash dtbo_a          boot_dtbo.img
 ./fastboot flash vbmeta_a        vbmeta.img
diff --git a/platform/tools/gensdimg.sh b/platform/tools/gensdimg.sh
index 3bdc604..7842ca0 100755
--- a/platform/tools/gensdimg.sh
+++ b/platform/tools/gensdimg.sh
@@ -74,6 +74,8 @@ gen_image() {
     ${PARTED_TOOL} add_partition ${BASE_ARGS} --partition-name=frp                   --size=512K
     ${PARTED_TOOL} add_image     ${BASE_ARGS} --partition-name=boot_a                --size=64M  --img-file=boot.img
     ${PARTED_TOOL} add_partition ${BASE_ARGS} --partition-name=boot_b                --size=64M
+    ${PARTED_TOOL} add_image     ${BASE_ARGS} --partition-name=init_boot_a           --size=8M   --img-file=init_boot.img
+    ${PARTED_TOOL} add_partition ${BASE_ARGS} --partition-name=init_boot_b           --size=8M
     ${PARTED_TOOL} add_image     ${BASE_ARGS} --partition-name=vendor_boot_a         --size=32M  --img-file=vendor_boot.img
     ${PARTED_TOOL} add_partition ${BASE_ARGS} --partition-name=vendor_boot_b         --size=32M
     ${PARTED_TOOL} add_image     ${BASE_ARGS} --partition-name=dtbo_a                --size=8M   --img-file=boot_dtbo.img
diff --git a/platform/tools/tools.mk b/platform/tools/tools.mk
index 8d152b0..b178ca2 100644
--- a/platform/tools/tools.mk
+++ b/platform/tools/tools.mk
@@ -25,6 +25,7 @@ DEPLOY_FILES := \
 	$(PRODUCT_OUT)/deploy-sd.img \
 	$(PRODUCT_OUT)/boot.img \
 	$(PRODUCT_OUT)/boot_dtbo.img \
+	$(PRODUCT_OUT)/init_boot.img \
 	$(PRODUCT_OUT)/vendor_boot.img \
 	$(PRODUCT_OUT)/super.img \
 	$(PRODUCT_OUT)/vbmeta.img \
diff --git a/platform/uboot/bootscript.cpp b/platform/uboot/bootscript.cpp
index 35f3367..cfeaf9f 100644
--- a/platform/uboot/bootscript.cpp
+++ b/platform/uboot/bootscript.cpp
@@ -32,6 +32,8 @@ EXTENV(partitions, ";name=misc,size=512K,uuid=\${uuid_gpt_misc}")
 EXTENV(partitions, ";name=frp,size=512K,uuid=\${uuid_gpt_frp}")
 EXTENV(partitions, ";name=boot_a,size=64M,uuid=\${uuid_gpt_boot_a}")
 EXTENV(partitions, ";name=boot_b,size=64M,uuid=\${uuid_gpt_boot_b}")
+EXTENV(partitions, ";name=init_boot_a,size=8M,uuid=\${uuid_gpt_init_boot_a}")
+EXTENV(partitions, ";name=init_boot_b,size=8M,uuid=\${uuid_gpt_init_boot_b}")
 EXTENV(partitions, ";name=vendor_boot_a,size=32M,uuid=\${uuid_gpt_vendor_boot_a}")
 EXTENV(partitions, ";name=vendor_boot_b,size=32M,uuid=\${uuid_gpt_vendor_boot_b}")
 EXTENV(partitions, ";name=dtbo_a,size=8M,uuid=\${uuid_gpt_dtbo_a}")
@@ -121,7 +123,7 @@ FUNC_BEGIN(bootcmd_start)
  then
   FEXTENV(bootargs, " androidboot.force_normal_boot=1") ;
  fi;
- abootimg addr \$loadaddr \$vloadaddr
+ abootimg addr \$loadaddr \$vloadaddr \$iloadaddr
 
  adtimg addr \${dtboaddr}
 #ifdef DEVICE_HANDLE_FDT
@@ -164,6 +166,9 @@ FUNC_BEGIN(bootcmd_block)
  part start mmc \$mmc_bootdev boot_\$slot_name boot_start &&
  part size  mmc \$mmc_bootdev boot_\$slot_name boot_size
 
+ part start mmc \$mmc_bootdev init_boot_\$slot_name init_boot_start &&
+ part size  mmc \$mmc_bootdev init_boot_\$slot_name init_boot_size
+
  part start mmc \$mmc_bootdev vendor_boot_\$slot_name vendor_boot_start &&
  part size  mmc \$mmc_bootdev vendor_boot_\$slot_name vendor_boot_size
 
@@ -172,6 +177,7 @@ FUNC_BEGIN(bootcmd_block)
 
  mmc dev \$mmc_bootdev &&
  mmc read \$loadaddr \$boot_start \$boot_size
+ mmc read \$iloadaddr \$init_boot_start \$init_boot_size
  mmc read \$vloadaddr \$vendor_boot_start \$vendor_boot_size
  mmc read \$dtboaddr \$dtbo_start \$dtbo_size
 FUNC_END()
diff --git a/platform/uboot/platform_amlogic.h b/platform/uboot/platform_amlogic.h
index 085cddf..b5e5b4a 100644
--- a/platform/uboot/platform_amlogic.h
+++ b/platform/uboot/platform_amlogic.h
@@ -3,7 +3,8 @@
 #define PLATFORM_SETUP_ENV() \
  setenv dtbaddr 0x5000000;   \
  setenv loadaddr 0x1000000;  \
- setenv vloadaddr 0x3000000; \
+ setenv vloadaddr 0x6000000; \
+ setenv iloadaddr 0x8000000; \
  setenv dtboaddr 0x4000000;  \
 
 #define PLATFORM_HANDLE_FDT() \
-- 
2.39.2

